Add PMC (PTP Management Client) collector

natifridman opened the pull request:
Add a collector to collect data from the PMC (PTP Management Client) command
```
pmc -u -f /var/run/ptp4l.0.config  'GET GRANDMASTER_SETTINGS_NP'
```


Collector output example:
```
*devices.PMCInfo:pmc-info, {"timeSource":"0xa0","clockAccuracy":"0xfe","offsetScaledLogVariance":"0xffff","clockClass":248,"currentUtcOffset":37,"leap61":0,"leap59":0,"currentUtcOffsetValid":0,"ptpTimescale":1,"timeTraceable":0,"frequencyTraceable":0}
```

crwr45 reviewed (changes requested):
Looks like you're on the right track. Please check the linter issues; you can run the linter locally with pre-commit (see the README). Please can you also reword your commits a bit to give a little more detail on what each one is doing and why.

natifridman commented on pkg/collectors/devices/pmc.go:
```diff
@@ -0,0 +1,58 @@
+	ClockAccuracy string `json:"ClockAccuracy" fetcherKey:"clockAccuracy"`
+}
+
+// AnalyserJSON returns the json expected by the analysers
```
Thanks, fixed

natifridman commented:
> Looks like you're on the right track. Please check the linter issues; you can run the linter locally with pre-commit (see the README). Please can you also reword your commits a bit to give a little more detail on what each one is doing and why.

Thanks, still need to fix one more linter issue.
I squashed the old commits and added to it a more meaningful message

nocturnalastro reviewed (changes requested):
Looking good just a few things to clean up.

natifridman commented on pkg/collectors/devices/pmc_test.go:
```diff
@@ -0,0 +1,82 @@
+				"<PMC>",
+				"sending: GET GRANDMASTER_SETTINGS_NP\n",
+				"507c6f.fffe.30fbe8-0 seq 0 RESPONSE MANAGEMENT GRANDMASTER_SETTINGS_NP\n",
+				"clockClass              248\n",
```
For some reason,  looks like it takes the `\n` as a text and not as a whitespace. Need to figure out why

natifridman replied:
Done

natifridman commented on pkg/collectors/devices/pmc.go:
```diff
@@ -0,0 +1,149 @@
+	FrequencyTraceable      int    `json:"frequencyTraceable"`
+}
+
+type GMSettings struct {
```
It was easier to use this when adding new properties

nocturnalastro replied:
I'm not sure what you mean here can you explain further?

natifridman replied:
When I needed to add something to the `PMCInfo` I had to add it only here and in the `processedResult`. Without the need to create another `fetcherKey`

nocturnalastro replied:
I don't think that's worth the nesting, your already adding the json key. It will also cause nesting in the output.

natifridman replied:
Done

nocturnalastro commented on pkg/collectors/devices/pmc.go:
```diff
@@ -0,0 +1,149 @@
+func (gmSetting *GMSettings) GetAnalyserFormat() ([]*callbacks.AnalyserFormatType, error) {
+	formatted := callbacks.AnalyserFormatType{
+		ID: "pmc",
+		Data: []any{
+			gmSetting.PMC,
+		},
```
It will also make the json output incorrect it should be values in an array or and an object (once my PR in vse-sync-pp gets merged. The current Will be an array with an object in it.

natifridman replied:
Could you please explain this?

natifridman commented on pkg/collectors/pmc_collector.go:
```diff
@@ -0,0 +1,105 @@
+type PMCCollector struct {
+	callback      callbacks.Callback
+	ctx           clients.ContainerContext
+	interfaceName string
```
Done

natifridman commented on pkg/collectors/devices/pmc.go:
```diff
@@ -0,0 +1,149 @@
+			`\s+timeTraceable\s+(\d+)\n` +
+			`\s+frequencyTraceable\s+(\d+)\n` +
+			`\s+timeSource\s+(.+)`,
+	)
```
Done

natifridman commented on pkg/collectors/devices/pmc.go:
```diff
@@ -0,0 +1,149 @@
+		return processedResult, fmt.Errorf("unable to parse pmc output: %s", result["PMC"])
+	}
+
+	clockClass, err := strconv.Atoi(match[1])
```
Done

nocturnalastro reviewed (changes requested):
Nice work just one last change to get a valid analyser format json.

natifridman commented on pkg/collectors/devices/pmc.go:
```diff
@@ -0,0 +1,148 @@
+func (gmSetting *PMCInfo) GetAnalyserFormat() ([]*callbacks.AnalyserFormatType, error) {
+	formatted := callbacks.AnalyserFormatType{
+		ID: "pmc",
+		Data: []any{
```
Thanks for the clarification. Done
